sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionWriter.scala

```diff
@@ -0,0 +1,137 @@
+import org.apache.spark.sql.execution.streaming.runtime.StreamingCheckpointConstants.DIR_NAME_STATE
+import org.apache.spark.sql.execution.streaming.state.{StateStore, StateStoreColFamilySchema, StateStoreConf, StateStoreId, StateStoreProvider, StateStoreProviderId}
+
+/**
+ * A writer that can directly write binary data to the streaming state store.
+ *
+ * This writer expects input rows with the same schema produced by
+ * StatePartitionAllColumnFamiliesReader:
+ *   (partition_key, key_bytes, value_bytes, column_family_name)
+ *
+ * The writer creates a fresh (empty) state store instance for the target commit version
+ * instead of loading previous partition data. After writing all rows for the partition, it will
+ * commit all changes as a snapshot
+ */
+class StatePartitionAllColumnFamiliesWriter(
+     storeConf: StateStoreConf,
+     hadoopConf: Configuration,
+     partitionId: Int,
+     targetCpLocation: String,
+     operatorId: Int,
+     storeName: String,
+     batchId: Long,
+     columnFamilyToSchemaMap: HashMap[String, StateStoreColFamilySchema]) {
+  private val defaultSchema = {
+    columnFamilyToSchemaMap.getOrElse(
+      StateStore.DEFAULT_COL_FAMILY_NAME,
+      throw new IllegalArgumentException(
+        s"Column family ${StateStore.DEFAULT_COL_FAMILY_NAME} not found in schema map")
+    )
+  }
+
+  private val columnFamilyToKeySchemaLenMap: MapView[String, Int] =
+    columnFamilyToSchemaMap.view.mapValues(_.keySchema.length)
+  private val columnFamilyToValueSchemaLenMap: MapView[String, Int] =
+    columnFamilyToSchemaMap.view.mapValues(_.valueSchema.length)
+
+  private val rowConverter = {
+    val schema = SchemaUtil.getScanAllColumnFamiliesSchema(defaultSchema.keySchema)
+    CatalystTypeConverters.createToCatalystConverter(schema)
+  }
+
+  protected lazy val provider: StateStoreProvider = {
+    val stateCheckpointLocation = new Path(targetCpLocation, DIR_NAME_STATE).toString
+    val stateStoreId = StateStoreId(stateCheckpointLocation,
+      operatorId, partitionId, storeName)
+    val stateStoreProviderId = StateStoreProviderId(stateStoreId, UUID.randomUUID())
+
+    val provider = StateStoreProvider.createAndInit(
+      stateStoreProviderId, defaultSchema.keySchema, defaultSchema.valueSchema,
+      defaultSchema.keyStateEncoderSpec.get,
+      useColumnFamilies = false, storeConf, hadoopConf,
+      useMultipleValuesPerKey = false, stateSchemaProvider = None)
+    provider
+  }
+
+  private lazy val stateStore: StateStore = {
+    // TODO[SPARK-54590]: Support checkpoint V2 in StatePartitionAllColumnFamiliesWriter
+    // Create empty store to avoid loading old partition data since we are rewriting the
+    // store e.g. during repartitioning
+    // Use loadEmpty=true to create a fresh state store without loading previous versions
+    // We create the empty store AT version, and the next commit will
+    // produce version + 1
+    provider.getStore(
+      batchId + 1,
```

Comment:
This is incorrect right. The current batch will load at `currentBatchId` and will commit `currentBatchId + 1` e.g. batch 0 will load 0 and commit 1.

This is different from the state reader, where the reader is asked to read the state produced by batchId, which is batchId + 1. e.g. to read the state for batch 0, you will need to load 1